### PR TITLE
Regrid dataset with dask variables

### LINKED ISSUE #121: Regridding dataset with dask arrays fails
MWE
```python3
import xesmf as xe

ds = xe.util.grid_global(1, 1)
ds = ds.assign(data=ds.lon + ds.lat, data2=ds.lon * ds.lat)

reg = xe.Regridder(ds, ds, 'bilinear')
reg(ds)  # ok

reg(ds.chunk())  # Fails
```
Fails with:
```
TypeError: Field elements must be 2- or 3-tuples, got 'dtype('float64')'
```
in `/dask/array/utils.py in meta_from_array(x, ndim, dtype)`

This is due to a misuse of `xr.apply_ufunc` in the dataset case.  The  `output_dtypes` argument of `apply_ufunc` is only used if data uses the dask backend or if `vectorize=True` (not our case).  It is a list of dtypes _the same length as the number of outputs_.  When encountering a dataset, `xarray` will iteratively apply the `ufunc` for each variable. Thus, in each case there is only 1 output. Passing a list to `output_dtypes` has no meaning.

The best fix would be that xarray could accept some sort of mapping? I'm not sure how complicated that would be, probably not trivial.

For xesmf, we could:
- Call `astype` ourselves on the output dataset.
- Go back to float output. 
- Midway : force float32 output when all input variables are float32, use float64 otherwise. 

(When regridding, integer dtypes don't really make sense, no?)

### PR DESCRIPTION
**Pull request by aulemahal**
Fixes #121

I chose to go with the easy-to-implement and covers-most-meaningful-cases option. Output will have the same dtype as the input EXCEPT when regridding dataset that have dask-backed variables. Once regridded, those variables will be coerced to 'float32' if and only if _all_ variables in the input dataset are 'float32', otherwise they will be casted as 'float64'.

### CODE REVIEW

**Review by huard — APPROVED** [review]

**Review by raphaeldussin — CHANGES_REQUESTED** [review]
this would undo the improvements from #102 
@brews opinion?

**Comment by aulemahal** [comment]
@raphaeldussin It shouldn't :  As I pointed out in #121, passing `output_dtypes` as a list doesn't have the effect it was thought to have. 
Explanation 1 : The output of `xarray.apply_ufunc` called on a Dataset is a Dataset. A Dataset is mapping, how should xarray interpret a `output_dtypes` list to cast the different _variables_ correctly? In some python versions, dictionaries have inconsistent order.

Explanation 2 : As stated in the doc string of `apply_ufunc` : 
> output_dtypes (list of dtype, optional) – Optional list of output dtypes. Only used if dask='parallelized' or vectorize=True. 
Looking at the code, I can further precise that the argument is used ONLY when applying the func on dask-backed variables.

Thus, my change do not affect the behaviour wanted in #102, i.e.: conserving dtypes on numpy-backed variables. The changes only affect lines that are ignored when passing numpy-backed variables.

Also, because of how xarray applies  `apply_ufunc` on Datasets and how it understands `output_dtypes`, there is no way for us to coerce the dtype of the different variables on the output. Except by calling `astype` ourselves _afterwards_, but that adds dask tasks and I don't feel it makes much sense (but I'm open to discuss).

**Comment by raphaeldussin** [comment]
@huard I will do my best to get to this by next week

File: xesmf/frontend.py [hunk 1]
```diff
@@ -555,7 +559,11 @@ def regrid_dataset(self, ds_in, keep_attrs=False, skipna=False, na_thres=1.0):
         ]
         ds_in = ds_in.drop_vars(non_regriddable)
 
-        ds_dtypes = [d.dtype for d in ds_in.data_vars.values()]
+        out_dtype = (  # Useful for dask-backed datasets
```
Comment (raphaeldussin) [inline xesmf/frontend.py:562]: wouldn't calling `.values()` trigger the loading into memory of the array??

I also think we could make this more generic with something like:

```python
dtypes = list(set(ds_in.data_vars.dtype))
if len(dtypes) == 1:
    out_dtype = dtypes[0]
else:
    size = 1  # init to 1 byte
    for d in dtypes:
        size = max(size, d.itemsize)  # use max number of bytes found in dtypes
    out_dtype = np.dtype(f"f{size}")
```

File: xesmf/tests/test_frontend.py [hunk 2]
```diff
@@ -566,6 +566,27 @@ def test_regrid_dataset():
...
+
+    # `ds_out` already refers to output grid object
+    ds_result = regridder(ds_in.chunk())
+
+    # output should contain all data variables
+    assert set(ds_result.data_vars.keys()) == set(ds_in.data_vars.keys())
+    assert dask.is_dask_collection(ds_result)
+    assert ds_result.data.dtype == ds_in.data.dtype
+
+    ds_in_32 = ds_in.copy()
```
Comment (raphaeldussin) [inline xesmf/tests/test_frontend.py:583]: although both are correct, I think people are more used to 4 bytes than 32 bits in the numerical types

File: xesmf/util.py [hunk 3]
```diff
@@ -109,7 +109,7 @@ def _flatten_poly_list(polys):
     """Iterator flattening MultiPolygons."""
     for i, poly in enumerate(polys):
         if isinstance(poly, MultiPolygon):
-            for sub_poly in poly:
+            for sub_poly in poly.geoms:
```
Comment (raphaeldussin) [inline xesmf/util.py:112]: relation to current PR?

**Review by raphaeldussin — CHANGES_REQUESTED** [review]

**Comment by raphaeldussin** [comment]
@aulemahal @huard I have just a few comments but I believe this can be dealt with quickly

File: xesmf/util.py (hunk 3 above)
Comment (huard) [reply xesmf/util.py:112]: I think this solves a DeprecationWarning.

**Comment by huard** [comment]
@raphaeldussin Ready for your re-review.

**Review by raphaeldussin — APPROVED** [review]

**Pull request merged**